### ISSUE TITLE
Error creating openstack_lb_loadbalancer_v2 with use_octavia="true"

### ISSUE THREAD

**Issue by hogepodge** [terraform-providers/terraform-provider-openstack#173]
### Terraform Version
11.0

### Affected Resource(s)
-  openstack_lb_loadbalancer_v2

### Terraform Configuration Files
```hcl
provider "openstack" {
  auth_url = "${ var.os_auth_url }"
  region = "${ var.os_region_name }"
  domain_name = "${ var.os_user_domain_name }"
  user_name = "${ var.os_username }"
  tenant_id = "${ var.os_project_id }"
  password = "${ var.os_password }"

  use_octavia = "${ var.use_octavia }"
}

resource "openstack_lb_loadbalancer_v2" "internal_lb" {
  name = "internal_lb"
  vip_subnet_id = "${ openstack_networking_subnet_v2.cncf.id }"
  vip_address = "${ var.internal_lb_ip }"
  security_group_ids = [ "${ var.security_group_id}" ]
} 
```

### Debug Output
http://paste.openstack.org/show/627664/

### Expected Behavior
A load balancer should have been created at the Octavia endpoint.

### Actual Behavior
The computed TenantID was not passed to the Octavia endpoint, resulting in "Validation failure: Missing project ID in request where one is required."

### Steps to Reproduce
1. obtain credential on Vexxhost
2. run terraform script that creates a v2 loadbalancer using octavia

### Important Factoids
Running in Vexxhost with an octavia specific endpoint. It's worth noting that the OpenStack LBaaS team strongly discourages using the Neutron LBaaS v2 API with automation tools such as Terraform and K8s. Use of the Neutron API when backed by Octavia (rather than the direct Octavia API) can result in race conditions that will leak ports owned by the Neutron service, resulting in ports, subnets, networks, and security groups that can not be deleted by an unprivileged user.

Changing the variable `use_octavia` to `false` results in successful load balancer creation.

**Comment by jtopjian** [comment]
It might have been fixed here:

https://github.com/terraform-providers/terraform-provider-openstack/pull/172

The above patch was merged a few days ago and will be available in the next release. You can compile this provider from source if you want to try it out early.

It's good to know that Vexxhost supports Octavia - I'm kind of running blind otherwise.

**Comment by jtopjian** [comment]
Quick note, if you want to compile from source, it's quite easy and only requires compiling _this_ plugin and not all of Terraform.

First install Go. [gimme](https://github.com/travis-ci/gimme) is my preferred method.

Next:

```shell
$ go get github.com/terraform-providers/terraform-provider-openstack
$ cd ~/go/src/github.com/terraform-providers/terraform-provider-openstack
$ make build
$ sudo mv terraform-provider-openstack /path/to/dir/where/terraform/is
$ cd /path/to/your/tf/file
$ terraform init
```

**Comment by hogepodge** [comment]
Octavia definitely doesn't accept "tenant_id" any more. Does Gopher Cloud support using "project_id" instead of "tenant_id"? This isn't just in the configuration, but in the messages sent to the OpenStack API.

https://review.openstack.org/#/c/462750/1/octavia/api/v2/controllers/load_balancer.py

**Comment by jtopjian** [comment]
> Octavia definitely doesn't accept "tenant_id"

OK, I think I'm starting to see what's going on here. I'm not sure of a solution yet, though. I might have to start poking at this first-hand.

Some background:

For v3 _authentication_, Gophercloud supports "project", though the field names are labeled as "tenant". It can be confusing. In short, Gophercloud should authenticate the user with the correct auth fields, though it might not seem obvious. If a service is checking your auth information, project will be available.

For _resources_ which can take a project or tenant, Gophercloud and Terraform use what is explicitly asked for. In the case of Neutron resources, they are currently all set to use `tenant_id`.  However, it's important to note that, so far, these have only been used for cases where an admin user is creating resources on behalf of other users.

From looking at the Terraform config you gave and the debug output, I'm a little confused. The server is sending you a message saying it wants a `project_id`. However, a `tenant_id` is not being sent in the request, so it's not like the error is a result of incorrect input. 

Why does Octavia want a `project_id` when creating a load balancer resource?

I'm only speculating, which might not be helpful, but possible reasons why could be:

* Octavia is doing different things than Neutron. We've only recently added Octavia "support". I say "support" because we're being clever piggybacking off of the Neutron LBaaS API and sending requests to the Octavia endpoint. Perhaps this isn't going to work.
* Vexxhost is doing something unique.
* Are you running this as an admin user by chance? According to the patch you gave, that should be the only time `project_id` is checked for:

```python
if context.is_admin or CONF.auth_strategy == constants.NOAUTH:
```

Of course, maybe there are other parts of Octavia that require a `project_id` field in the resource creation request. Like I mentioned before: I'm running blind at the moment. :(

/cc @jrperritt

**Comment by jtopjian** [comment]
I just reviewed the patch and story @dims posted. I'm not sure if that's exactly what you're running into, but it seems to be similar:

https://review.openstack.org/#/c/523889/2/octavia/api/v2/controllers/load_balancer.py

Looking at the original code, though, `context.project_id` _should_ be available. That's from the token information, correct?

**Comment by hogepodge** [comment]
I'm entirely sure myself. Vexxhost did some downstream patching to perhaps set the context. This speaks to a larger problem with LBaaS. The service projects (be it Neutron or Octavia) are creating service-scope resources on non-privileged project-scope networks. More precisely, in the case of using the Neutron LBaaSv2 API, Neutron-owned ports are being created on user project-owned subnets.

When the OpenStack Terraform provider tears down the load balancer, it can trigger a race condition in Neutron that puts the load balancer into an error state with dangling Neutron-owned ports. Neutron is unable to do the cascading delete, and the user can not remove any resource not cleaned up by Neutron because they don't have service-level credentials.

The Octavia team has suggested that direct API access to Octavia will mitigate this issue, as the direct access does not depend on polling Neutron for resource updates. I'm trying to express all of this to the best of my knowledge and abilities, and don't necessarily understand everything that's happening under the hood.

tl;dr, direct Octavia access should be the way to go. I'll test with the downstream changes and see if that makes a difference. We could try to get you credential on Vexxhost if that would help in debugging too.

Thanks for your patience in helping me to understand what's going on.

**Comment by jtopjian** [comment]
Thanks for the info, Chris.

> We could try to get you credential on Vexxhost if that would help in debugging too.

That's currently in progress, actually 😄

**Comment by xgerman** [comment]
@jtopjian all Octavia resources if used by an admin for another tenant require project_id (https://developer.openstack.org/api-ref/load-balancer/v2/index.html) Though dims request will fix it I am skeptical that it will get approved so we need to fix that in gopher cloud. I have been using a user tenant and not set project_id/tenant_id and things seem fine so far.

@all I discovered another problem when assigning a FloatingIP which occurs if you are not an admin. Fix has been approved for master and is being backported to Pike (https://review.openstack.org/#/q/0f4a5e21709c77c926cbbf7929217fa951b48ad7)

**Comment by hogepodge** [comment]
Quick update on this, with the downstream changes the load balancer gets created, but the API is sending back a 405 which causes Terraform to halt and not manage the resource. Good news is I can manage the resources manually and clean out my tenant if things go wrong. http://paste.openstack.org/show/627987/

**Comment by jtopjian** [comment]
@xgerman 

> I have been using a user tenant and not set project_id/tenant_id and things seem fine so far.

So you are authenticating as a non-admin user with project_name/tenant_name scope?

**Comment by xgerman** [comment]
@jtopjian Yes, that's correct.

@hogepodge I have been told that the event_streamer will change PENDING->ERROR if it's working. Because of the way we set up rabbitMQ that never worked. Though I made a fix ( https://review.openstack.org/#/c/514452/) we decided to go with the Octavia API endpoint. So I can't report if it's working or not for sure.

**Comment by mnaser** [comment]
I think I have an idea of what's going wrong.  Given that the resource is created successfully, we know that this code path is completed:

https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_lb_loadbalancer_v2.go#L126-L137

However, I think what is going wrong is here:

https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_lb_loadbalancer_v2.go#L139-L147

However, the VIP port is not owned by the tenant when using the Octavia API until recently, which meant that you were trying to update a resource you do not own.  This fix was added 3 days ago.

https://review.openstack.org/#/c/524254/

I will sync up our Octavia to the latest stable/pike at VEXXHOST and update everyone.  @xgerman mentioned that it fixes things.

**Comment by mnaser** [comment]
We've deployed latest stable/pike in addition to this patch https://review.openstack.org/#/c/523901/ (pending merge).  I believe that this should solve this issue (probably good to know for anyone else using Octavia).

@hogepodge could you have a look?

**Comment by jtopjian** [comment]
Hi all,

I've just tested this on VEXXHOST and can confirm that the problem is resolved -- I was able to successfully create a load balancer stack of the following:

```hcl
provider "openstack" {
  use_octavia = "true"
}

resource "openstack_networking_network_v2" "network_1" {
  name = "network_1"
  admin_state_up = "true"
}

resource "openstack_networking_subnet_v2" "subnet_1" {
  name = "subnet_1"
  cidr = "192.168.199.0/24"
  ip_version = 4
  network_id = "${openstack_networking_network_v2.network_1.id}"
}

resource "openstack_lb_loadbalancer_v2" "loadbalancer_1" {
  name = "loadbalancer_1"
  vip_subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
}

resource "openstack_lb_listener_v2" "listener_1" {
  name = "listener_1"
  protocol = "HTTP"
  protocol_port = 8080
  loadbalancer_id = "${openstack_lb_loadbalancer_v2.loadbalancer_1.id}"
}

resource "openstack_lb_pool_v2" "pool_1" {
  name = "pool_1"
  protocol = "HTTP"
  lb_method = "ROUND_ROBIN"
  listener_id = "${openstack_lb_listener_v2.listener_1.id}"
}

resource "openstack_lb_member_v2" "member_1" {
  address = "192.168.199.10"
  protocol_port = 8080
  weight = 10
  admin_state_up = "true"
  pool_id = "${openstack_lb_pool_v2.pool_1.id}"
  subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
}

resource "openstack_lb_member_v2" "member_2" {
  address = "192.168.199.11"
  protocol_port = 8080
  weight = 15
  admin_state_up = "true"
  pool_id = "${openstack_lb_pool_v2.pool_1.id}"
  subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
}
```

I'm going to go ahead and consider this issue fixed and close it. However, I'm happy to re-open if there is still a problem here.

@mnaser Thank you for your help with this. I'm going to be running some of the Terraform OpenStack acceptance tests on VEXXHOST. What's the best way of contacting you or someone else if I run into any issues?
